src/librustc/mir/visit.rs

```diff
@@ -880,125 +945,156 @@ pub enum TyContext {
+}
+
 impl<'tcx> PlaceContext<'tcx> {
-    /// Returns true if this place context represents a drop.
+    /// Returns `true` if this place context represents a drop.
     pub fn is_drop(&self) -> bool {
         match *self {
-            PlaceContext::Drop => true,
+            PlaceContext::MutatingUse(MutatingUseContext::Drop) => true,
             _ => false,
         }
     }
 
-    /// Returns true if this place context represents a storage live or storage dead marker.
+    /// Returns `true` if this place context represents a storage live or storage dead marker.
     pub fn is_storage_marker(&self) -> bool {
         match *self {
-            PlaceContext::StorageLive | PlaceContext::StorageDead => true,
+            PlaceContext::NonUse(NonUseContext::StorageLive) |
+            PlaceContext::NonUse(NonUseContext::StorageDead) => true,
             _ => false,
         }
     }
 
-    /// Returns true if this place context represents a storage live marker.
+    /// Returns `true` if this place context represents a storage live marker.
     pub fn is_storage_live_marker(&self) -> bool {
         match *self {
-            PlaceContext::StorageLive => true,
+            PlaceContext::NonUse(NonUseContext::StorageLive) => true,
             _ => false,
         }
     }
 
-    /// Returns true if this place context represents a storage dead marker.
+    /// Returns `true` if this place context represents a storage dead marker.
     pub fn is_storage_dead_marker(&self) -> bool {
         match *self {
-            PlaceContext::StorageDead => true,
+            PlaceContext::NonUse(NonUseContext::StorageDead) => true,
             _ => false,
         }
     }
 
-    /// Returns true if this place context represents a use that potentially changes the value.
+    /// Returns `true` if this place context represents a use that potentially changes the value.
     pub fn is_mutating_use(&self) -> bool {
         match *self {
-            PlaceContext::Store | PlaceContext::AsmOutput | PlaceContext::Call |
-            PlaceContext::Borrow { kind: BorrowKind::Mut { .. }, .. } |
-            PlaceContext::Projection(Mutability::Mut) |
-            PlaceContext::Drop => true,
-
-            PlaceContext::Inspect |
-            PlaceContext::Borrow { kind: BorrowKind::Shared, .. } |
-            PlaceContext::Borrow { kind: BorrowKind::Shallow, .. } |
-            PlaceContext::Borrow { kind: BorrowKind::Unique, .. } |
-            PlaceContext::Projection(Mutability::Not) |
-            PlaceContext::Copy | PlaceContext::Move |
-            PlaceContext::StorageLive | PlaceContext::StorageDead |
-            PlaceContext::Validate => false,
+            PlaceContext::MutatingUse(..) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Borrow {
+                kind: BorrowKind::Mut { .. },
+                ..
+            }) |
+            PlaceContext::MaybeMutatingUse(MaybeMutatingUseContext::Projection(Mutability::Mut)) |
```

Comment:
Now we still have two lists of classifications, if only for the `MaybeMutatingUse`. The entire point was to avoid doing that so that *by construction*, it is impossible for a `PlaceContext` to be both a mutating and non-mutating use. Your approach does not achieve that.

Reply:
Going to resolve this and respond in the other review comment.